### PR TITLE
Coot 14, Remove `arma` from `arma::zeros` and `arma::ones`

### PR DESCRIPTION
**Pull request by shrit**

### CODE REVIEW

File: src/mlpack/methods/ann/not_adapted/brnn_impl.hpp [hunk 1]
```diff
@@ -384,10 +384,10 @@ EvaluateWithGradient(const arma::mat& /* parameters */,
...
   if (backwardGradient.is_empty())
   {
-    backwardGradient = arma::zeros<arma::mat>(
+    backwardGradient = zeros<arma::mat>(
         parameter.n_rows/ 2,
         parameter.n_cols);
-    forwardGradient = arma::zeros<arma::mat>(
+    forwardGradient = zeros<arma::mat>(
         parameter.n_rows/ 2,
         parameter.n_cols);
```
Comment (rcurtin) [inline src/mlpack/methods/ann/not_adapted/brnn_impl.hpp:392]: Both of these (`backwardGradient` and `forwardGradient`) can fit on one line now. :+1:

File: src/mlpack/methods/hmm/hmm_impl.hpp [hunk 2]
```diff
@@ -257,8 +257,8 @@ void HMM<Distribution>::Train(const std::vector<arma::mat>& dataSeq,
         << ")." << std::endl;
   }
 
-  arma::mat initial = arma::zeros(logInitial.n_elem);
-  arma::mat transition = arma::zeros(logTransition.n_rows,
+  arma::mat initial = zeros(logInitial.n_elem);
+  arma::mat transition = zeros(logTransition.n_rows,
                                      logTransition.n_cols);
```
Comment (rcurtin) [inline src/mlpack/methods/hmm/hmm_impl.hpp:262]: This will fit on one line too.

File: src/mlpack/methods/local_coordinate_coding/lcc_impl.hpp [hunk 3]
```diff
@@ -184,7 +184,7 @@ inline void LocalCoordinateCoding::OptimizeDictionary(
 
   // Build dataPrime := [X x^1 ... x^1 ... x^n ... x^n]
   // where each x^i is repeated for the number of neighbors x^i has.
-  arma::mat dataPrime = arma::zeros(data.n_rows,
+  arma::mat dataPrime = zeros(data.n_rows,
```
Comment (rcurtin) [inline src/mlpack/methods/local_coordinate_coding/lcc_impl.hpp:187]: This fits on one line now. :+1:

File: src/mlpack/methods/lsh/lsh_search_impl.hpp [hunk 4]
```diff
@@ -554,9 +554,9 @@ void LSHSearch<SortPolicy, MatType>::GetAdditionalProbingBins(
   // Actions vector describes what perturbation (-1/+1) corresponds to a score.
   arma::Col<short int> actions(2 * numProj); // will be [-1 ... 1 ...]
   actions.rows(0, numProj - 1) = // First numProj rows.
-    -1 * arma::ones< arma::Col<short int> > (numProj); // -1s
+    -1 * ones< arma::Col<short int> > (numProj); // -1s
```
Comment (rcurtin) [inline src/mlpack/methods/lsh/lsh_search_impl.hpp:557]:
```suggestion
    -1 * ones<arma::Col<short int>> (numProj); // -1s
```
This was written so long ago that you couldn't put two closing `>>`s next to each other :smile:

File: src/mlpack/methods/lsh/lsh_search_impl.hpp [hunk 5]
```diff
@@ -554,9 +554,9 @@ void LSHSearch<SortPolicy, MatType>::GetAdditionalProbingBins(
...
   actions.rows(numProj, (2 * numProj) - 1) = // Last numProj rows.
-    arma::ones< arma::Col<short int> > (numProj); // 1s
+    ones< arma::Col<short int> > (numProj); // 1s
```
Comment (rcurtin) [inline src/mlpack/methods/lsh/lsh_search_impl.hpp:559]:
```suggestion
    ones<arma::Col<short int>> (numProj); // 1s
```

File: src/mlpack/methods/rann/ra_search_rules_impl.hpp [hunk 6]
```diff
@@ -60,7 +60,7 @@ RASearchRules(const arma::mat& referenceSet,
   numSamplesReqd = RAUtil::MinimumSamplesReqd(n, k, tau, alpha);
 
   // Initialize some statistics to be collected during the search.
-  numSamplesMade = arma::zeros<arma::Col<size_t> >(querySet.n_cols);
+  numSamplesMade = zeros<arma::Col<size_t> >(querySet.n_cols);
```
Comment (rcurtin) [inline src/mlpack/methods/rann/ra_search_rules_impl.hpp:63]:
```suggestion
  numSamplesMade = zeros<arma::Col<size_t>>(querySet.n_cols);
```
Another one :)

**Review by rcurtin — COMMENTED** [review]
Looks good to me, just the usual smattering of style comments :smile:

File: HISTORY.md [hunk 7]
```diff
@@ -21,8 +21,22 @@
...
+  * Remove `arma` from `arma::function`, (#3619), (#3617) (#3610), (#3643), (#3600),
+    (#3605), (#3629).
+
+  * Change `arma::conv_to` to `ConvTo` using a local shim (#3614).
+
+  * Fix a bug for the stddev and mean in `RandNormal()` #(3651).
+
   * Allow PCA to take different matrix types (#3677).
 
+>>>>>>> master
```
Comment (rcurtin) [inline HISTORY.md:39]: Oops, still a merge artifact here?

File: src/mlpack/methods/ann/not_adapted/brnn_impl.hpp [hunk 8]
```diff
@@ -384,12 +384,8 @@ EvaluateWithGradient(const arma::mat& /* parameters */,
...
   if (backwardGradient.is_empty())
   {
-    backwardGradient = arma::zeros<arma::mat>(
-        parameter.n_rows/ 2,
-        parameter.n_cols);
-    forwardGradient = arma::zeros<arma::mat>(
-        parameter.n_rows/ 2,
-        parameter.n_cols);
+    backwardGradient = zeros<arma::mat>(parameter.n_rows/ 2, parameter.n_cols);
+    forwardGradient = zeros<arma::mat>(parameter.n_rows/ 2, parameter.n_cols);
```
Comment (rcurtin) [inline src/mlpack/methods/ann/not_adapted/brnn_impl.hpp:388]:
```suggestion
    backwardGradient = zeros<arma::mat>(parameter.n_rows / 2, parameter.n_cols);
    forwardGradient = zeros<arma::mat>(parameter.n_rows / 2, parameter.n_cols);
```
(I did check that it still fits in 80 characters even with the space :))

**Review by rcurtin — APPROVED** [review]
Thanks, the cleanups are much appreciated.  I might suggest trying to add a bit of the "why" to the history, but, up to you.  Someone reading things like "Remove `arma`" is probably not going to understand why, but a message more like "Generalize matrix operations in preparation for GPU support" gets the idea across to someone who has no idea what's going on internally.  But, your call :+1:

File: HISTORY.md (hunk 7 above)
Comment (shrit) [reply HISTORY.md:39]: Oops, this indeed is a merge artifact,
Will fix that shortly, and include the above message

**Pull request merged**